pkg/compactor/block_upload.go

```diff
@@ -99,11 +104,16 @@ func (c *MultitenantCompactor) FinishBlockUpload(w http.ResponseWriter, r *http.
 		return
 	}
 
-	if err := c.completeBlockUpload(ctx, logger, userBkt, blockID, *m); err != nil {
-		writeBlockUploadError(err, op, "", logger, w)
+	// create validation file
+	if err := c.uploadValidation(ctx, logger, blockID, userBkt); err != nil {
+		writeBlockUploadError(err, op, "while creating validation file", logger, w)
 		return
 	}
 
+	if c.compactorCfg.EnableBackgroundValidation {
```

Comment:
Isn't it a bug that block completion is only done if background validation is enabled?

If I'm not wrong, shouldn't block completion conditionally start background validation?

Reply:
I am not sure I follow...

Validation will always be started, unless it's explicitly disabled by this flag, which is only set by unit test to avoid starting background goroutine. If validation is disabled, block will never be completed correctly. But that will only be possible in the test code.

Reply:
@pstibrany never mind. Sorry, I was thinking of partial block cleanup and the parameter for enabling it (since I just re-reviewed that PR). I do find it confusing though that the parameter implies disabling of background validation, but also disables actually completing the block. If you're not aware of the context surrounding it (i.e. unit tests as you say), it's confusing for the reader.